File: server/apps/nexotype/schemas/user/user_biomarker_reading_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class UserBiomarkerReadingCreate(BaseModel):
    """Schema for creating a new user biomarker reading"""
    subject_id: int
    biomarker_id: int
    source_id: int
    value: float
    unit_id: int
    measured_at: datetime


class UserBiomarkerReadingUpdate(BaseModel):
    """Schema for updating a user biomarker reading"""
    subject_id: int = None
    biomarker_id: int = None
    source_id: int = None
    value: float = None
    unit_id: int = None
    measured_at: datetime = None


# ==========================================
# Response Schemas
# ==========================================

class UserBiomarkerReadingDetail(BaseModel):
    """Schema for user biomarker reading details"""
    id: int
    subject_id: int
    biomarker_id: int
    source_id: int
    value: float
    unit_id: int
    measured_at: datetime
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class UserBiomarkerReadingListResponse(BaseModel):
    """Response schema for listing user biomarker readings"""
    success: bool
    data: list[UserBiomarkerReadingDetail] | None = None
    count: int = 0
    error: str | None = None


class UserBiomarkerReadingResponse(BaseModel):
    """Response schema for single user biomarker reading operations"""
    success: bool
    data: UserBiomarkerReadingDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/user/user-biomarker-reading.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const UserBiomarkerReadingSchema = z.object({
  id: z.number(),
  subject_id: z.number().int(),
  biomarker_id: z.number().int(),
  source_id: z.number().int(),
  value: z.number(),
  unit_id: z.number().int(),
  measured_at: z.string(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateUserBiomarkerReadingSchema = z.object({
  subject_id: z.number().int(),
  biomarker_id: z.number().int(),
  source_id: z.number().int(),
  value: z.number(),
  unit_id: z.number().int(),
  measured_at: z.string().min(1),
});

export const UpdateUserBiomarkerReadingSchema = z.object({
  subject_id: z.number().int().optional(),
  biomarker_id: z.number().int().optional(),
  source_id: z.number().int().optional(),
  value: z.number().optional(),
  unit_id: z.number().int().optional(),
  measured_at: z.string().optional(),
});

// ==========================================
// Types
// ==========================================

export type UserBiomarkerReading = z.infer<typeof UserBiomarkerReadingSchema>;
export type CreateUserBiomarkerReading = z.infer<typeof CreateUserBiomarkerReadingSchema>;
export type UpdateUserBiomarkerReading = z.infer<typeof UpdateUserBiomarkerReadingSchema>;

// ==========================================
// Response Types
// ==========================================

export type UserBiomarkerReadingResponse = {
  success: boolean;
  data?: UserBiomarkerReading;
  error?: string;
};

export type UserBiomarkerReadingsResponse = {
  success: boolean;
  data?: UserBiomarkerReading[];
  count?: number;
  error?: string;
};
